lib/web/recordingplayback.go

```diff
@@ -0,0 +1,767 @@
+				data:        websocket.FormatCloseMessage(websocket.CloseUnsupportedData, "only binary messages are supported"),
+			}:
+			case <-time.After(1 * time.Second):
+				s.logger.ErrorContext(s.ctx, "timeout sending close message", "session_id", s.sessionID)
+			}
+
+			return
+		}
+
+		req, err := decodeBinaryRequest(data)
+		if err != nil {
+			s.logger.WarnContext(s.ctx, "failed to decode request", "session_id", s.sessionID, "error", err)
+			continue
+		}
+
+		switch req.requestType {
+		case requestTypeFetch:
+			s.handleFetchRequest(req)
+		default:
+			s.sendError(trace.BadParameter("unknown request type: %d", req.requestType), req.requestID)
+
+			s.logger.ErrorContext(s.ctx, "received unknown request type", "session_id", s.sessionID, "type", req.requestType)
+		}
+	}
+}
+
+// createTaskContext creates a new context for a task and cancels any previous task.
+// A task context is used to manage the lifecycle of a fetch request, ensuring that only one fetch request is active at a time.
+func (s *recordingPlayback) createTaskContext() context.Context {
+	s.mu.Lock()
+
+	if s.cancelActiveTask != nil {
+		// Cancel the active task first
+		s.cancelActiveTask()
+		s.mu.Unlock()
+
+		// Wait for streamEvents to terminate before continuing
+		// We unlock the mutex while waiting to avoid deadlock
+		s.wg.Wait()
+
+		s.mu.Lock()
+	}
+
+	ctx, taskCancel := context.WithCancel(s.ctx)
+	s.cancelActiveTask = taskCancel
+	s.mu.Unlock()
+
+	return ctx
+}
+
+// handleFetchRequest processes a fetch request for session events.
+func (s *recordingPlayback) handleFetchRequest(req *fetchRequest) {
+	if err := validateRequest(req); err != nil {
+		s.sendError(err, req.requestID)
+
+		return
+	}
+
+	ctx := s.createTaskContext()
+
+	s.stream.Lock()
+
+	// start the stream if it doesn't exist or if we need to go back in time
+	needNewStream := s.stream.eventsChan == nil || req.startOffset < s.stream.lastEndTime
+
+	if needNewStream {
+		events, errors := s.clt.StreamSessionEvents(
+			metadata.WithSessionRecordingFormatContext(ctx, teleport.PTY),
+			session.ID(s.sessionID),
+			0,
+		)
+
+		if events == nil || errors == nil {
+			s.sendError(fmt.Errorf("failed to start session event stream"), req.requestID)
+			s.stream.Unlock()
+
+			return
+		}
+
+		s.stream.eventsChan = events
+		s.stream.errorsChan = errors
+		s.stream.lastEndTime = 0
+
+		s.terminal.Lock()
+		s.terminal.vt = vt10x.New()
+		s.terminal.Unlock()
+	}
+
+	s.stream.lastEndTime = req.endOffset
+
+	eventsChan := s.stream.eventsChan
+	errorsChan := s.stream.errorsChan
+
+	s.stream.Unlock()
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.streamEvents(ctx, req, eventsChan, errorsChan)
+	}()
+}
+
+// streamEvents streams session events to the client.
+func (s *recordingPlayback) streamEvents(ctx context.Context, req *fetchRequest, eventsChan <-chan apievents.AuditEvent, errorsChan <-chan error) {
+
+	startSent := false
+	screenSent := false
+	inTimeRange := false
+	var streamStartTime time.Time
```

Comment:
Why is the purpose of streamStartTime  ? 

it is set when empty to `streamStartTime = evt.GetTime()` but never used later.